### DISCUSSION TITLE
Using react-hook-form with a Date/Time Picker and using Material-UI?

### DISCUSSION THREAD

**Discussion by tkentdna** [react-hook-form/react-hook-form#3769] [Q&A]
Is there an example of using react-hook-form with a Date/Time Picker and using Material-UI? I have been able to implement with a Mui TextField that has a type of "datetime-local", and I can set the Date/Time via manual interaction with the control, but I have been unable to get a timestamp to appear in the field upon entering the form using default values for react-hook-form; that is, the date/time value doesn't appear in the picker. Upon manually setting the date/time value and submitting the form, the value is getting properly bound to the react-hook-form "data" object. Some fragments from my code are below. I have replaced irrelevant code with ellipses (...).

`
import React from "react";
import { Controller, useForm } from "react-hook-form";
import { TextField } from "@material-ui/core";
    ...

    interface IMyFormProps {
      name: string;
      description: string;
      occurrenceTimestamp: Date;
    }

    ...

    const MyForm: React.FC<IMyFormProps> = (props: IMyFormProps) => {

      ...

      // set up details for ReactHookForm
      const { register, control, formState, handleSubmit, errors } = useForm({
        defaultValues: {
          name: props.name,
          description: props.description,
          occurrenceTimestamp: props.occurrenceTimestamp
        },
        mode: "all"
      });

      ...

      return (
        <form onSubmit=... 
            // a text input for Name
            <TextField
              inputRef={register({ required: true })}
              name="name"
              label="Name"
              ...
            />

            // a text input for Description
            <TextField
              inputRef={register({ required: true })}
              name="description"
              label="Description"
              ...
            />

            // The Date/Time Picker
            <Controller
              render={(props) =>
                <TextField
                  inputRef={register({ required: true })}
                  {...props}
                  type="datetime-local"
                  label="Occurrence Date/Time"
                />
              }
              name="occurrenceTimestamp"
              control={control}
            >
            </Controller>

        >
        </form>
      )
    }
`

**Comment by bluebill1049** [comment]
have you seen this example: https://codesandbox.io/s/react-hook-form-v6-controller-qsd8r

**Reply by tkentdna** [reply]
Thanks! I'll review the example for ideas.

**Reply by tkentdna** [reply]
Bill, thanks again for pointing me to the example.  It turns out that I hadn't included a proper onChange handler (onChange={(e) => props.onChange(e)}) and that made the difference.  With that, I experimented with a few different DateTime pickers and ended up choosing the DateTimePicker and KeyboardDateTimePicker published by the Material-UI group (in the **@material-ui/pickers** package) as the controls provide solid functionality and nice theming out of the box.

**Reply by ajjack50n** [reply]
Hey .. I would be interested in your solution. I am trying to use DateTimePicker with React-hook-form

**Reply by McGern** [reply]
@tkentdna I am also interested in your solution. We had the keyboarddatetimepicker working with a combo of MUI, rhf and yup, but when entering with the keyboard input, the mask breaks and it starts throwing "cannot read property length of undefined". Just trying to get to which part of our setup is breaking it, so would be great to have a working demo.

**Reply by ajjack50n** [reply]
@McGern
So I eventually got got this working. I think for me it was all about getting the onChange event setup correctly - I tried a number of things. This is what I hit on in the end...

I had to use MomentUtils...
It was important to set the `defaultValues` and then populate the form using `reset`

and then I created the following component...

```
import React from 'react'
import { Controller } from 'react-hook-form'
import { KeyboardDatePicker } from "@material-ui/pickers"

const SmartDatePicker = ({ name, label, control, setValue }) => {

    return (
        <Controller
            name={name}
            control={control}
            setValue={setValue}
            render={({ onChange, value, name }) => (
                <KeyboardDatePicker margin="normal" disableToolbar variant="inline" inputVariant="outlined" label={label} format="YYYY-MM-DD" name={name} value={value}
                    minDate={new Date('1400-01-01')} autoOk disableFuture
                    maxDateMessage="Future dates are not allowed" minDateMessage="Date must be after 1400"
                    onChange={(e, date) => setValue(name, date)}
                />
            )}
        />
    )
}
export default SmartDatePicker

```
I struggled with this for a few days!!!  but the above now works for me. Only ongoing issue, which I have accepted, is that I get a warning if I edit the textbox  using the keyboard,  saying momentUtils is rolling back to Date because non ISO format used. Which, of course, is true until I finish editing. I would like to not get the warning but, I guess, it could be worse. Before I was having the same problems as you.

**Reply by ak7550** [reply]
@ajjack50n 
Even I am struggling with the same issue. My code is constantly giving me:
**Unhandled Rejection (TypeError): Cannot Read the Property 'value' of undefined**

I also have set the  `defaultValue` for the dob part only, and tried to copy your instructions accordingly, but  there is some issue still there. 
Here is my code:
```javascript
const { register, handleSubmit, formState: { errors }, reset, clearErrors, watch, control, setValue } = useForm({
        mode: "onBlur",
        reValidateMode: "onBlur",
        criteriaMode: "all",
        defaultValues: {
            dob: new Date(),
        }
    });

const onSubmit = (unserInfo, event) => {
        console.log(unserInfo);
        console.log(`submit button clicked`);
     reset();
    }

    const onError = (errors, event) => {
        console.log("error is: ", errors);
        handleError(errors);
        reset();
        clearErrors();
    }
return (
        <form onSubmit={handleSubmit(onSubmit, onError)}>
         <Controller
                                    control={control}
                                    name="dob"
                                    setValue={setValue}
                                    placeholderText="Select your birth date"
                                    onChange={date => setValue(date)}
                                    render={({ onChange, value, name }) => (
                                        <MuiPickersUtilsProvider utils={DateFnsUtils}>
                                            <KeyboardDatePicker
                                                format="dd/MM/yyyy"
                                                margin="normal"
                                                id="date-picker-inline"
                                                inputVariant="outlined"
                                                variant="inline"
                                                label="dob"
                                                name={name}
                                                value={value}
                                                minDate={new Date('1400-01-01')}
                                                minDateMessage="Date must be after 1400"
                                                maxDateMessage="Future dates are not allowed"
                                                autoOk
                                                disableFuture
                                                onChange={(e, date)=> setValue(name, date)}
                                                KeyboardButtonProps={{
                                                    'aria-label': 'change date',
                                                }}
                                                {
                                                ...register("dob")
                                                }
                                                className={classes.radiogroup}
                                                style={{
                                                    marginTop: '0.1em'
                                                }}
                                            />
                                        </MuiPickersUtilsProvider>
                                    )}
                                />    
         </form>
);
```
I am complete beginner in react-hook-form and trying to learn it from scratch. There might be some silly errors please pardon me for the same. 

It would be grateful for me if you help me to find solution.
Here is the screenshot of my errors:

![Screenshot (126)](https://user-images.githubusercontent.com/54152266/117539428-350a9b00-b028-11eb-92dd-e887fcd56d4e.png)
![Screenshot (127)](https://user-images.githubusercontent.com/54152266/117539429-38058b80-b028-11eb-8135-eaff14f92a21.png)

**Comment by McGern** [comment]
Thanks very much for providing that. We have something similar, but like
you we are getting errors when using the keyboard entry. Using datefns for
the date utilities, don't know if that makes a difference.

**Comment by ognitio-technologies** [comment]
I am using material ui 5 and react-hook-form 7, showing console error while implementing date picker
```

` <LocalizationProvider dateAdapter={AdapterDayjs}>
          <Controller
            control={control}
            name="feedbackDate"
            defaultValue={new Date()}
            render={({ field }) => (
              <DesktopDatePicker
                {...field}
                label="Feedback date"
                value={field.value}
                onChange={(e) => field.onChange(e)}
                renderInput={(params) => <TextField {...params} fullWidth />}
              />
            )}
          />
        </LocalizationProvider>`
```

Error: 
![Screenshot 2022-04-23 at 3 24 17 PM](https://user-images.githubusercontent.com/72864567/164889655-8dc24db3-9b63-4ec2-98f4-6a64165a250c.png)

**Reply by Moshyfawn** [reply]
You're passing the RHF `field.ref` callback into the `DatePicker`'s `ref` prop by spreading `{...field}` into the component. You want to pass the RHF `field.ref` callback into `DatePicker`'s `inputRef` prop instead
```diff
 <Controller
-  render={({ field }) => (
+  render={({ field: { ref, ...fieldProps } }) => (
     <DesktopDatePicker
-      {...field}
+      {...fieldProps}
+      inputRef={ref}
     />
   )}
 />
```

**Reply by ognitio-technologies** [reply]
> You're passing the RHF `field.ref` callback into the `DatePicker`'s `ref` prop by spreading `{...field}` into the component. You want to pass the RHF `field.ref` callback into `DatePicker`'s `inputRef` prop instead
> 
> ```diff
>  <Controller
> -  render={({ field }) => (
> +  render={({ field: { ref, ...fieldProps } }) => (
>      <DesktopDatePicker
> -      {...field}
> +      {...fieldProps}
> +      inputRef={ref}
>      />
>    )}
>  />
> ```

Thanks working
